clang-tools-extra/clang-tidy/daedalean/IncludeOrderCheck.cpp

```diff
@@ -0,0 +1,193 @@
+//===--- IncludeOrderCheck.cpp - clang-tidy -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "IncludeOrderCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/PPCallbacks.h"
+#include "clang/Lex/Preprocessor.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace daedalean {
+
+namespace {
+
+enum class GroupType {
+  RelatedHeader = 0,
+  SameDirectory,
+  SameProject,
+  External
+};
+
+class IncludeOrderPPCallbacks : public PPCallbacks {
+public:
+  explicit IncludeOrderPPCallbacks(ClangTidyCheck &Check,
+                                   const SourceManager &SM)
+      : Check(Check), SM(SM) {}
+
+  void InclusionDirective(SourceLocation HashLoc, const Token &IncludeTok,
+                          StringRef FileName, bool IsAngled,
+                          CharSourceRange FilenameRange, const FileEntry *File,
+                          StringRef SearchPath, StringRef RelativePath,
+                          const Module *Imported,
+                          SrcMgr::CharacteristicKind FileType) override;
+  void EndOfMainFile() override;
+
+private:
+  struct IncludeDirective {
+    SourceLocation Loc;
+    CharSourceRange Range;
+    std::string Filename;
+
+    GroupType groupType;
+    std::string project;
+
+    bool operator <(const IncludeDirective & other) const {
+      if (groupType != other.groupType) {
+        return static_cast<int>(groupType) < static_cast<int>(other.groupType);
+      }
+      return  Filename < other.Filename;
+    }
+  };
+
+  typedef std::vector<IncludeDirective> FileIncludes;
+  std::map<clang::FileID, FileIncludes> IncludeDirectives;
+
+  ClangTidyCheck &Check;
+  const SourceManager &SM;
+
+private:
+  void printInverse(const IncludeDirective & what, const IncludeDirective & beforeWhat);
+};
+
+std::string basename(const std::string & path) {
+  if (const size_t pos = path.find_last_of('/'); pos == std::string::npos) {
+    return {};
```

Comment:
I'd argue that it should return `path` in this case; the basename of "bla.c" is "bla.c"